### PR TITLE
[gen_l10n] keys can contain dollar sign

### LINKED ISSUE #112250: [gen_l10n] A name that has any dollar sign ($) isn't allowed.
flutter_localizations(gen_l10n comamnd) doesn't allow a property name to have a dollar sign ($) in any postion.

## Steps to Reproduce

1. Execute `flutter gen-l10n` involved in any *.arb file which has a key that contains a dollar sign ($).
2. Loot at the terminal output.

**Expected results:** Dollar sign can be a part of the property name, so it should succeed.

**Actual results:** `gen-l10n` failed.

## Code sample
In  app_en.arb file:
``` json
{
  "$address": "Address:"
}
```

## Logs
```
Invalid ARB resource name "$address" in LocalFile: '<ProjectFolder>\lib/l10n\app_en.arb'.
Resources names must be valid Dart method names: they have to be camel case, cannot start with a number or
underscore, and cannot contain non-alphanumeric characters.
```

### PR DESCRIPTION
**Pull request by asashour**
The l10n tool currently rejects keys with a dollar sign, however Dart allows it to be used in methods and getters, so there is no need to prevent it from being used.

This PR changed the generator to accept the dollar sign as a valid character in the name. 

Fixes #112250

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by christopherfujino** [comment]
@thkim1011 can you review this one?

**Review by thkim1011 — APPROVED** [review]
LGTM!

File: packages/flutter_tools/lib/src/localizations/gen_l10n.dart [hunk 1]
```diff
@@ -825,14 +825,14 @@ class LocalizationsGenerator {
       return false;
     }
     // Dart getter and method name cannot contain non-alphanumeric symbols
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/localizations/gen_l10n.dart:827]:
```suggestion
    // Dart identifiers can only use letters, numbers, underscores, and `$`
```

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**